src/ConnectionTrait.php

```diff
@@ -118,20 +122,17 @@ public function resetAttemptCount(): void
     /**
      * @param string $connectionName
      */
-    public function connect($connectionName = null)
+    public function connect(?string $connectionName = null): DriverConnection
     {
         $this->hasBeenClosedWithAnOpenTransaction = false;
 
         /** @psalm-suppress InternalMethod */
         return parent::connect($connectionName);
     }
 
-    public function close()
+    public function close(): void
     {
-        if ($this->getTransactionNestingLevel() > 0) {
-            $this->hasBeenClosedWithAnOpenTransaction = true;
-        }
-
+        $this->hasBeenClosedWithAnOpenTransaction = $this->getTransactionNestingLevel() > 1;
```

Comment:
checking for nested transaction using `0` is not correct, because "main" transaction is actually `1`, so nested will be 1+

Reply:
We're not checking for a nested transaction here, but for an open transaction at all. If `close` is being called with a transaction level different from zero, that means that the transaction didn't get committed properly and we encountered and exception, so we may not know if the requested query (which may have been a write query) has been received or not by the server. Hence, down below we do 
```php
    public function canTryAgain(\Throwable $throwable, string $sql = null): bool
    {
        if ($this->hasBeenClosedWithAnOpenTransaction) {
            return false;
        }
```

I'm reverting this.

Reply:
Done in 0f1119e84eb77c3c8f625a7df3888619c63837a2 + 4fbc26aacd15d48176df34b233c4f273c5c6f379